autodistill/core/custom_detection_model.py

```diff
@@ -26,6 +40,26 @@ def predict(self, image: str) -> sv.Detections:
 
         detections = self.detection_model.predict(image)
 
+        if self.set_of_marks is not None:
+            labels = [f"{num}" for num in range(len(detections.xyxy))]
+
+            opened_image = np.array(opened_image)
+
+            annotated_frame = self.set_of_marks_annotator.annotate(
+                scene=opened_image, labels=labels, detections=detections
+            )
+
+            opened_image = Image.fromarray(annotated_frame)
+
+            opened_image.save("temp.jpeg")
+
+            result = self.classification_model.predict("temp.jpeg")
```

Comment:
I don't quite follow how this works. Is `result` going to be a single label even if there are multiple objects?

Reply:
Set of mark only makes sense when you use a custom prompt, like this:

```python
SAMGPT = CustomDetectionModel(
    detection_model=GroundedSAM(
        CaptionOntology({"apple": "apple", "mango": "mango"})
    ),
    classification_model=GPT4V(
        CaptionOntology({k: k for k in classes}),
        api_key="KEY",
        prompt=f"Assign a label to each object. Return a response in NUMBER: LABEL, new line, etc. You can use the classes {', '.join(classes)}. If a class does not match, say 'None'.",
    ),
    set_of_mark=True
)
```

There will be a `set_of_mark_detection` function in GPT-4V that runs Set of Marks with the custom prompt, then generates an sv.Detections object with the results.

The "classification_model" bit here is confusing because technically GPT-4V is doing Set of Mark, but the `classification_model` that you would use is GPT. We will need to document that this only makes sense to use with LMMs (and GPT is the only model that will implement Set of Mark in Autodistill right now).

Reply:
The overall flow is:

- SAM generates masks.
- The `set_of_mark` flag, when enabled, takes the source image and applies SoM annotations.
- A user uses an LMM (right now, only GPT-4V) with a custom prompt. This triggers a special function in the base model that handles SoM (the results logic will vary between each model due to the different inference code required to get results from each model).

Reply:
What happens if the LLM doesn't precisely follow the directions? (eg gives a response in a different format, adds extra descriptions, skips a mark, etc?)

We should probably be able to verify whether `classification_model` supports set of marks. Would that mean inheriting from `SetOfMarksModel` & checking the type if you pass `True` for `set_of_marks`? Or is there a better way to validate that?

Are you forced to use the same `CaptionOntology` here?

Reply:
> What happens if the LLM doesn't precisely follow the directions? (eg gives a response in a different format, adds extra descriptions, skips a mark, etc?)

This will be up to the model, and may vary over time (i.e. if GPT Vision gets JSON mode, that may be preferable to the text output, but for now we ask for text output). We will write the requisite code to process the LLM output (i.e. strip unnecessary spaces, only save predictions for class in the ontology) for each model.

I have added in a validation check: if a base model has a `set_of_mark` function, you can use Set of Mark. We will only add this in GPT Vision since that is where the research points. Should other models perform well with SoM, we can add support, too.

A `SetOfMarksModel` class doesn't seem like a good idea because Set of Mark is a prompting technique, not a model itself. SoM could be used with other LMMs in the future. Inheriting from another class confuses the class architecture of `autodistill`. We can validate by checking if a base model has an function attr that we can use to retrieve results.